File: adventurejs/lib/parse_steps/tokenize.js
const errors = require('../errors/tokenization.js');
const matcher = require('../grammar_matcher.js');

module.exports = context => {
  let expectedBlockTerminator = null;
  let unterminatedBlock = null;

  for(let instruction of context.instructions) {

    if(unterminatedBlock) {
      const match = expectedBlockTerminator.exec(instruction.line);

      if(match) {
        instruction.name = match[2];
        instruction.type = 'BLOCK_TERMINATOR';

        const dashes = match[1];
        const operatorColumn = instruction.line.indexOf(dashes);
        const nameColumn = instruction.line.lastIndexOf(instruction.name);

        instruction.ranges = {
          blockOperator: [operatorColumn, operatorColumn + dashes.length],
          name: [nameColumn, nameColumn + instruction.name.length]
        };

        expectedBlockTerminator = null;
        unterminatedBlock = null;

        continue;
      }

      instruction.ranges = { content: [0, instruction.line.length] };
      instruction.type = 'BLOCK_CONTENT';

      continue;
    }

    const match = matcher.GRAMMAR_REGEXP.exec(instruction.line);

    if(match) {

      if(match[matcher.EMPTY_INDEX] !== undefined) {
        instruction.type = 'EMPTY';
        continue;
      }


      if(match[matcher.COMMENT_ANGLE_INDEX]) {
        instruction.type = 'COMMENT';

        const operatorColumn = instruction.line.indexOf('>');
        instruction.ranges = { commentOperator: [operatorColumn, operatorColumn + 1] };

        instruction.text = match[matcher.COMMENT_TEXT_INDEX] || null;

        if(instruction.text) {
          const textColumn = instruction.line.lastIndexOf(instruction.text);
          instruction.ranges.comment = [textColumn, textColumn + instruction.text.length];
        }

        continue;
      }


      if(match[matcher.NAME_OPERATOR_INDEX]) {
        instruction.name = match[matcher.NAME_UNESCAPED_INDEX] ||
                           match[matcher.NAME_ESCAPED_INDEX];

        const nameColumn = instruction.line.indexOf(instruction.name);
        const operatorColumn = instruction.line.indexOf(':', nameColumn + instruction.name.length);

        instruction.ranges = {
          nameOperator: [operatorColumn, operatorColumn + 1],
          name: [nameColumn, nameColumn + instruction.name.length]
        };

        const value = match[matcher.FIELD_VALUE_INDEX];
        if(value) {
          instruction.type = 'FIELD';
          instruction.value = value;

          const valueColumn = instruction.line.lastIndexOf(value);
          instruction.ranges.value = [valueColumn, valueColumn + value.length];
        } else {
          instruction.type = 'NAME';
          instruction.ranges.value = [operatorColumn + 1, instruction.line.length];
        }

        continue;
      }


      if(match[matcher.LIST_ITEM_INDEX]) {
        instruction.type = 'LIST_ITEM';
        instruction.value = match[matcher.LIST_ITEM_VALUE_INDEX] || null;

        const operatorColumn = instruction.line.indexOf('-');
        instruction.ranges = { itemOperator: [operatorColumn, operatorColumn + 1] };

        if(instruction.value) {
          const valueColumn = instruction.line.lastIndexOf(instruction.value);
          instruction.ranges.value = [valueColumn, valueColumn + instruction.value.length];
        } else {
          instruction.ranges.value = [operatorColumn + 1, instruction.line.length];
        }

        continue;
      }


      if(match[matcher.DICTIONARY_ENTRY_EQUALS_INDEX]) {
        instruction.name = match[matcher.NAME_UNESCAPED_INDEX] ||
                    match[matcher.NAME_ESCAPED_INDEX];

        instruction.type = 'DICTIONARY_ENTRY';
        instruction.value = match[matcher.DICTIONARY_ENTRY_VALUE_INDEX] || null;

        const nameColumn = instruction.line.indexOf(instruction.name); // TODO: Account for ``` `` ``` possibility ... :)
        const operatorColumn = instruction.line.indexOf('=', nameColumn + instruction.name.length);

        instruction.ranges = {
          entryOperator: [operatorColumn, operatorColumn + 1],
          name: [nameColumn, nameColumn + instruction.name.length]
        };

        if(instruction.value) {
          const valueColumn = instruction.line.lastIndexOf(instruction.value);
          instruction.ranges.value = [valueColumn, valueColumn + instruction.value.length];
        } else {
          instruction.ranges.value = [operatorColumn + 1, instruction.line.length];
        }

        continue;
      }


      const blockDashes = match[matcher.BLOCK_DASHES_INDEX];
      if(blockDashes) {
        instruction.name = match[matcher.BLOCK_NAME_INDEX];
        instruction.type = 'BLOCK';

        const operatorColumn = instruction.line.indexOf(blockDashes);
        const nameColumn = instruction.line.lastIndexOf(instruction.name);
        instruction.ranges = {
          blockOperator: [operatorColumn, operatorColumn + blockDashes.length],
          name: [nameColumn, nameColumn + instruction.name.length]
        };

        const nameEscaped = instruction.name.replace(/[-/\\^$*+?.()|[\]{}]/g, '\\$&');
        expectedBlockTerminator = new RegExp(`^\\s*(${blockDashes})\\s*(${nameEscaped})\\s*$`);
        unterminatedBlock = instruction;

        continue;
      }


      if(match[matcher.APPEND_WITH_NEWLINE_OPERATOR_INDEX]) {
        instruction.separator = '\n';
        instruction.type = 'FIELD_APPEND';
        instruction.value = match[matcher.APPEND_WITH_NEWLINE_VALUE_INDEX] || null;

        const pipeColumn = instruction.line.indexOf('|');
        instruction.ranges = { pipe: [pipeColumn, pipeColumn + 1] };

        if(instruction.value) {
          const valueColumn = instruction.line.lastIndexOf(instruction.value);
          instruction.ranges.value = [valueColumn, valueColumn + instruction.value.length];
        } else {
          instruction.ranges.value = [pipeColumn + 1, instruction.line.length];
        }

        continue;
      }


      if(match[matcher.APPEND_WITH_SPACE_OPERATOR_INDEX]) {
        instruction.separator = ' ';
        instruction.type = 'FIELD_APPEND';
        instruction.value = match[matcher.APPEND_WITH_SPACE_VALUE_INDEX] || null;

        const backslashColumn = instruction.line.indexOf('\\');
        instruction.ranges = { backslash: [backslashColumn, backslashColumn + 1] };

        if(instruction.value) {
          const valueColumn = instruction.line.lastIndexOf(instruction.value);
          instruction.ranges.value = [valueColumn, valueColumn + instruction.value.length];
        } else {
          instruction.ranges.value = [backslashColumn + 1, instruction.line.length];
        }

        continue;
      }


      const sectionHashes = match[matcher.SECTION_HASHES_INDEX];
      if(sectionHashes) {
        instruction.depth = sectionHashes.length;
        instruction.name = match[matcher.SECTION_NAME_UNESCAPED_INDEX] ||
                           match[matcher.SECTION_NAME_ESCAPED_INDEX];

        instruction.type = 'SECTION';

        const hashesColumn = instruction.line.indexOf(blockDashes);
        const nameColumn = instruction.line.indexOf(instruction.name, hashesColumn + sectionHashes.length);

        instruction.ranges = {
          hashes: [hashesColumn, hashesColumn + sectionHashes.length],
          name: [nameColumn, nameColumn + instruction.name.length]
        };

        const template = match[matcher.SECTION_TEMPLATE_INDEX];
        if(template) {
          instruction.template = template;

          const copyColumn = instruction.line.indexOf('<', nameColumn + instruction.name.length);
          const templateColumn = instruction.line.indexOf(template, copyColumn + 1);

          instruction.ranges.copyOperator = [copyColumn, copyColumn + 1];
          instruction.ranges.template = [templateColumn, templateColumn + template.length];
        }

        continue;
      }


      const template = match[matcher.TEMPLATE_INDEX];
      if(template) {
        instruction.name = match[matcher.NAME_UNESCAPED_INDEX] ||
                           match[matcher.NAME_ESCAPED_INDEX];

        instruction.template = template;
        instruction.type = 'NAME';

        const nameColumn = instruction.line.indexOf(instruction.name);
        const operatorColumn = instruction.line.indexOf('<', nameColumn + instruction.name.length);
        const templateColumn = instruction.line.indexOf(template, operatorColumn + 1);

        instruction.ranges = {
          template: [templateColumn, templateColumn + template.length],
          copyOperator: [operatorColumn, operatorColumn + 1],
          name: [nameColumn, nameColumn + instruction.name.length]
        };

        continue;
      }

    } else {
      errors.invalidLine(context, instruction);
    }
  }

  if(unterminatedBlock) {
    errors.unterminatedBlock(context, unterminatedBlock);
  }
};

File: enorb/lib/eno/tokenize.rb
module Eno
  class Tokenizer
    def initialize(input)
      @input = input
      @char = input[0]
      @column = 1
      @line = 1
      @index = 0
      @instructions = []
    end

    def block
      instruction = {
        line: @line,
        ranges: {
          block_operator: [@column, @column + 1]
        },
        type: :block
      }

      while next! == '-'
        instruction[:ranges][:block_operator][1] += 1
      end

      whitespace

      token = last_token

      if token
        instruction[:ranges][:name] = token[:range]
        instruction[:name] = token[:value]
      else
        ERROR
      end

      @instructions << instruction

      next!

      # while !terminator_line read content line instruction

      # read terminator instruction
    end

    def comment
      instruction = {
        line: @line,
        ranges: {
          comment_operator: [@column, @column + 1]
        },
        type: :comment
      }

      next!

      whitespace

      token = last_token

      if token
        instruction[:ranges][:text] = token[:range]
        instruction[:text] = token[:value]
      end

      @instructions << instruction
    end

    def empty
      @instructions << {
        line: @line,
        type: :empty
      }
    end

    def field
      next! while @char != "\n"
    end

    def last_token
      begin_column = @column
      begin_index = @index
      end_column = @column
      end_index = @index

      begin
        break if @char == "\n"
        if @char != ' '
          end_column = @column
          end_index = @index
        end
      end while next!

      if begin_index == end_index
        nil
      else
        {
          range: [begin_column, end_column],
          value: @input[begin_index..end_index]
        }
      end
    end

    def list_item
      next! while @char != "\n"
    end

    def name
      instruction = {
        ranges: {}
      }

      begin_column = @column
      begin_index = @index
      end_column = @column
      end_index = @index

      begin
        break if @char == "<"
        break if @char == ":"
        break if @char == "="
        raise 'ERROR' if @char == "\n"
        if @char != ' '
          end_column = @column
          end_index = @index
        end
      end while next!

      case @char
      when ":"
        instruction[:ranges][:name] = [begin_column, end_column]
        instruction[:ranges][:name_operator] = [@column, @column + 1]

        whitespace

        if @char == "\n"
          instruction[:type] = :name
        else
          field
        end
      when "="
        instruction[:ranges][:name] = [begin_column, end_column]
        instruction[:ranges][:entry_operator] = [@column, @column + 1]

        whitespace

        if @char == "\n"
          instruction[:type] = :name
        else
          field
        end
      end

      next! while @char != "\n"
    end

    def next!
      if @char == "\n"
        @column = 1
        @line += 1
      else
        @column += 1
      end

      @index += 1

      putc @input[@index] unless @input[@index] == nil

      @char = @input[@index]
    end

    def section
      instruction = {
        depth: 1,
        line: @line,
        ranges: {
          section_operator: [@column, @column + 1]
        },
        type: :section
      }

      while next! == '#'
        instruction[:depth] += 1
        instruction[:ranges][:section_operator][1] += 1
      end

      whitespace

      case @char
      when "\n"
        raise 'ERROR'
      when '`'
        escaped_name
      else
        token = last_token

        instruction[:name] = token[:value]
        instruction[:ranges][:name] = token[:range]
      end

      @instructions << instruction
    end

    def tokenize
      begin
        whitespace

        case @char
        when "\n"
          empty
        when '>'
          comment
        when '`'
          escaped_name
        when '#'
          section
        when '-'
          if @input[@index + 1] == '-'
            block
          else
            list_item
          end
        else
          name
        end
      end while next!

      puts @instructions
    end

    def whitespace
      next! while @char == ' '
    end
  end
end
